File: http-servers/typescript/deno-oak/src/database/repository.ts
import { env } from "../config/env.ts";
import { CassandraUserRepository } from "./cassandra.ts";
import { MongoUserRepository } from "./mongodb.ts";
import { PostgresUserRepository } from "./postgres.ts";
import { RedisUserRepository } from "./redis.ts";
import type { CreateUser, UpdateUser, User } from "./types.ts";

export interface UserRepository {
  create(data: CreateUser): Promise<User>;
  findById(id: string): Promise<User | null>;
  update(id: string, data: UpdateUser): Promise<User | null>;
  delete(id: string): Promise<boolean>;
  deleteAll(): Promise<void>;
  healthCheck(): Promise<boolean>;
  disconnect(): Promise<void>;
}

export const databaseTypes = [
  "postgres",
  "mongodb",
  "redis",
  "cassandra",
] as const;
export type DatabaseType = (typeof databaseTypes)[number];

const repositories = new Map<DatabaseType, UserRepository>();

const repositoryFactories: Record<DatabaseType, () => UserRepository> = {
  postgres: () => new PostgresUserRepository(env.POSTGRES_URL),
  mongodb: () => new MongoUserRepository(env.MONGODB_URL, env.MONGODB_DB),
  redis: () => new RedisUserRepository(env.REDIS_URL),
  cassandra: () =>
    new CassandraUserRepository({
      contactPoints: env.CASSANDRA_CONTACT_POINTS,
      localDataCenter: env.CASSANDRA_LOCAL_DATACENTER,
      keyspace: env.CASSANDRA_KEYSPACE,
    }),
};

export function isDatabaseType(value: string): value is DatabaseType {
  return databaseTypes.includes(value as DatabaseType);
}

export function getRepository(database: DatabaseType): UserRepository {
  let repo = repositories.get(database);
  if (!repo) {
    repo = repositoryFactories[database]();
    repositories.set(database, repo);
  }
  return repo;
}

export function resolveRepository(database: string): UserRepository | null {
  if (!isDatabaseType(database)) return null;
  return getRepository(database);
}

File: http-servers/python/fastapi/src/database/repository.py
from __future__ import annotations

import asyncio
from typing import Literal, Protocol

from src.config.env import env
from src.database.types import CreateUser, UpdateUser, User

DatabaseType = Literal["postgres", "mongodb", "redis", "cassandra"]
DATABASE_TYPES: list[DatabaseType] = ["postgres", "mongodb", "redis", "cassandra"]


class UserRepository(Protocol):
    async def create(self, data: CreateUser) -> User: ...
    async def find_by_id(self, id: str) -> User | None: ...
    async def update(self, id: str, data: UpdateUser) -> User | None: ...
    async def delete(self, id: str) -> bool: ...
    async def delete_all(self) -> None: ...
    async def health_check(self) -> bool: ...
    async def disconnect(self) -> None: ...


_repositories: dict[DatabaseType, UserRepository] = {}


def get_repository(database: DatabaseType) -> UserRepository:
    if database in _repositories:
        return _repositories[database]

    repo: UserRepository
    if database == "postgres":
        from src.database.postgres import PostgresUserRepository

        repo = PostgresUserRepository(env.POSTGRES_URL)
    elif database == "mongodb":
        from src.database.mongodb import MongoUserRepository

        repo = MongoUserRepository(env.MONGODB_URL, env.MONGODB_DB)
    elif database == "redis":
        from src.database.redis_repo import RedisUserRepository

        repo = RedisUserRepository(env.REDIS_URL)
    elif database == "cassandra":
        from src.database.cassandra import CassandraUserRepository

        repo = CassandraUserRepository(
            env.CASSANDRA_CONTACT_POINTS, env.CASSANDRA_LOCAL_DATACENTER, env.CASSANDRA_KEYSPACE
        )
    else:
        raise ValueError(f"Unknown database type: {database}")

    _repositories[database] = repo
    return repo


def resolve_repository(database: str) -> UserRepository | None:
    if database not in DATABASE_TYPES:
        return None
    return get_repository(database)  # type: ignore[arg-type]


async def initialize_databases() -> None:
    await asyncio.gather(*[get_repository(db).health_check() for db in DATABASE_TYPES])


async def disconnect_databases() -> None:
    for repo in _repositories.values():
        await repo.disconnect()
    _repositories.clear()
